### ISSUE TITLE
[Bug]: EnvWorker deadlock/zombie subprocess in single-GPU IsaacLab setup (GR00T N1.6)

### ISSUE THREAD

**Issue by salmadaoud0-cloud** [RLinf/RLinf#1384]
### Affected component

rlinf/envs/isaaclab, rlinf/scheduler/collective, GR00T N1.6 embodied training pipeline

### Bug description

When running RLinf's embodied training pipeline with the IsaacLab environment
on a single-GPU setup, the training hangs indefinitely with no error message.

Two related symptoms observed:
1. A child process under `EnvWorker` becomes a zombie (`<defunct>`, confirmed
   via `ps aux`), while the parent `EnvWorker` remains alive, blocked
   indefinitely on `queue.get()` waiting for a response that never comes
   (stack trace: `rlinf/envs/isaaclab/venv.py:99` → `reset()` →
   `_init_isaaclab_env` → `multiprocessing.queues.get`).

2. In a separate run, a deadlock between `CollectiveManager` and
   `EmbodiedFSDPActor.recv_rollout_trajectories`, confirmed via `py-spy dump`
   on both processes:
   - `EmbodiedFSDPActor` stuck in `_atomic_recv` (collective_group.py:464) →
     `_init_process_group` (collective_group.py:1217) → `_init_group`
     (collective_group.py:1188) → `_get_group_info_safe` (collective.py:89)
     → `ray.get(...)` that never returns.
   - `CollectiveManager` actor itself is idle (`main_loop`, empty task
     queue) — the request appears to be lost or never dispatched.
   - Meanwhile `MultiStepRolloutWorker` has a separate thread already past
     initialization, stuck instead on `_broadcast`/`recv` waiting for actual
     tensor data — suggesting each side is waiting for the other to act
     first (weight_syncer init_sync vs. rollout trajectory collection).

Reproduced with two different configs:
- Our custom IsaacLab task (VialsToRack, SO-101 robot, GR00T N1.6)
- Also reproduced with the official example config
  (`isaaclab_franka_stack_cube_ppo_gr00t.yaml`), adapted only for
  single-GPU placement (`cluster.component_placement` set to `0` instead
  of `0-3`, `total_num_envs` reduced to 2) — confirming this is not
  specific to our custom task/environment.

Already tried without effect:
- Forcing `NCCL_SOCKET_IFNAME=lo`, `GLOO_SOCKET_IFNAME=lo`,
  `NCCL_P2P_DISABLE=1` (raw NCCL init for FSDP completes fine — the hang is
  at RLinf's own collective layer, not NCCL/PyTorch distributed itself)
- Disabling `weight_syncer.patch.init_sync.enabled` (no change in symptom)

Note: `TIMEOUT_WARN_TIME` in `cluster.py` is set to 3600000 (effectively
1 hour via a 1ms sleep busy-wait loop in `_get_group_info_safe`), so this
hang produces no warning log within any reasonable test duration, making it
silent and hard to diagnose. Might be worth lowering for debug/dev use.

Question: is single-GPU "collocated" placement (env/rollout/actor all on
GPU 0) fully supported for the IsaacLab + GR00T embodied training path?

### Configuration YAML file

cat /workspace/RLinf/examples/embodiment/config/isaaclab_ppo_gr00t_vials_to_rack.yaml

### Log file

LATEST_LOG=$(ls -t /workspace/RLinf/logs/manual-*/run.log | head -1)
tail -200 "$LATEST_LOG"

### Environment

python -V
python -m pip list | grep -iE "torch|transformers|flash-attn|rlinf|ray"
nvidia-smi

### Other reproduction info

RLinf version: cloned from main branch, mid-July 2026
Docker image tag: custom (teleop-docker:with-rlinf-final, based on Isaac Sim container)
This was reproduced on a fresh RTX 6000 Ada machine after migrating from an
RTX 3080 (16GB) where we initially hit CUDA OOM errors — those were resolved
by moving to a GPU with more VRAM, but this deadlock persisted independently
of GPU model. Happy to share full config/log files or run additional
diagnostics (py-spy dumps, etc.) if useful for reproduction.

### Before submitting a new issue...

- [x] Have you checked relevant issues, FAQs (https://rlinf.readthedocs.io/en/latest/rst_source/resources/faq.html), or asked the chatbot at the top right corner of the [documentation page](https://rlinf.readthedocs.io).

**Comment by zoeyuchao** [comment]
@guozhen1997 please help with this issue. Thx!

**Comment by yangzhongii** [comment]
@salmadaoud0-cloud HI, I am the author of RLinf Gr00t N1.6. Let me assist you. As I have checked the codes, I find that it is not the problems of Gr00t N1,6 computation layers, it is the problems of isaaclab env. But don't you worry, I can help to fix the bugs.

**Comment by yangzhongii** [comment]
Hi @salmadaoud0-cloud , I've investigated both symptoms:

---

## Bug 1: EnvWorker zombie / queue.get() hang

**Confirmed and fixed.** The root cause is in `rlinf/envs/isaaclab/venv.py`:

- `env_fn()` (which launches IsaacSim) was called **outside** the `try:` block (line 33). If IsaacSim startup failed, the child process died silently.
- `obs_queue.get()` (lines 99, 108) had **no timeout** — parent blocked forever waiting for a dead child.
- `close()` (line 112) would raise `BrokenPipeError` if the child was already dead.

Fix (PR-ready):
- Moved `env_fn()` inside the `try:` block; catch all exceptions instead of only `KeyboardInterrupt`
- Added `timeout=300` to `queue.get()` with child-aliveness checks
- Guarded `close()` against `BrokenPipeError` / `EOFError`, added `join(timeout=10)`
- Added `_check_child_alive()` helper for early detection of child process death
- All changes are purely defensive — zero impact on normal execution paths.

---

## Bug 2: CollectiveManager deadlock

**We could not reproduce this.** Test setup:

| What | Our setup |
|---|---|
| Config | `isaaclab_franka_stack_cube_ppo_gr00t.yaml` |
| Placement | `env,rollout,actor: 0` (single-GPU collocated) |
| Model | GR00T N1.5 (`model/gr00t`, same config you used) |
| Docker image | `rlinf/rlinf:agentic-rlinf0.3-isaaclab` (official) |
| GPU | RTX 4090 (single GPU) |
| RLinf version | main branch, late July 2026 |

Training completed multiple rollout epochs and steps — collective communication initialized normally, weight sync worked, no hang. We also tested GR00T N1.6 + LIBERO on single-GPU collocated (`actor,env,rollout: all`) with the same result.

Additionally, we reduced `TIMEOUT_WARN_TIME` in `cluster.py` from 3,600,000 to 60,000 and added a hard `TimeoutError` after 10 minutes in both `_get_worker_info_safe` and `_get_group_info_safe` in `collective.py`. Any future hang in the collective layer will now produce a clear error within minutes instead of being silent for an hour.

---

## Request

Since the deadlock didn't reproduce with the official Docker image, it may be triggered by your custom environment. Could you share:

1. **Your Docker image** — `teleop-docker:with-rlinf-final`, or its Dockerfile
2. **Your custom config** — `isaaclab_ppo_gr00t_vials_to_rack.yaml`
3. **Full run logs** from both failing configs

---

## Workaround

For anyone hitting Bug 1 right now, the fix is available at:
https://github.com/yangzhongii/RLinf/tree/fix/issue-1384-deadlock-zombie


The `TIMEOUT_WARN_TIME` change and hard timeout in `collective.py` touch the collective communication layer shared by all models and environments. The architects should review the exact timeout values and failure behavior before landing on `main`.

**Comment by yangzhongii** [comment]
@guozhen1997 Could you please review the branch below? It touches `rlinf/scheduler/collective/` and `rlinf/scheduler/cluster/` — the timeout changes affect all embodied pipelines. If it looks good, I'll open a PR.

https://github.com/yangzhongii/RLinf/tree/fix/issue-1384-deadlock-zombie

Summary of changes:
- `rlinf/envs/isaaclab/venv.py`: defensive error handling for IsaacSim subprocess (Bug 1)
- `rlinf/scheduler/cluster/cluster.py`: `TIMEOUT_WARN_TIME` 3600000 → 60000
- `rlinf/scheduler/collective/collective.py`: hard `TimeoutError` after 10 min in busy-wait loops (Bug 2)
